### ISSUE TITLE
provenance/v1-rc1: Ability to store inline byproducts

### ISSUE THREAD

**Issue by lcarva** [slsa-framework/slsa#649]
In version 0.2 of SLSA Provenance, the unstructured nature of `buildConfig` allowed build systems to store additional information about the build that didn't necessarily fit neatly under other attributes. [Tekton Chains](https://github.com/tektoncd/chains/) leverages this when generating SLSA Provenance v0.2 for builds generated by a [Tekton Pipeline](https://tekton.dev/docs/pipelines/). 

Today Chains includes Task results in the SLSA Provenance v0.2 attestation. (A Pipeline is composed of Tasks, and Tasks can generate a result.) These results are small. Having them in the attestation means a verifier can leverage their value during a policy evaluation. 

For example, I have a Pipeline that includes a Task that performs certain checks against the checked out code, and produces a mini report, e.g. `{"x": 5, "y": 0, "z": 99}`. Once the build is complete, I evaluate its provenance attestation against a policy that looks at those values to decide whether or not the underlying build artifact should be promoted to the next environment.

Having the ability to specify an Artifact, not just an ArtifactReference in byproducts would allow me to continue to leverage this workflow.

The workaround is to push those results to an external location, and simply use the existing ArtifactReference. However, this has a few downsides. First, infrastructure to store such results is required. Second, verification becomes more complex.

**Comment by MarkLodato** [comment]
I'm not opposed to this for cases where you really do want to add just a tiny bit of extra information. Maybe a `value` field that is an arbitrary JSON object, probably mutually exclusive with `digest`?

@marcelamelara FYI since you're probably interested and might want to weigh in

**Comment by MarkLodato** [comment]
> A URI wouldn't be needed either since the artifact info is in-line, meaning that it's no longer a reference, I think. Could we retrofit the `mediaType` field to indicate that the object should be parsed as an in-line object by the consumer?

I think these are separate concepts:

- What the artifact represents: `uri`
- How it was obtained: `downloadLocation` (or `uri`)
- Its value: `digest`, `value` (if accepted)
- How the value was interpreted: `mediaType`, `localName`

**Comment by marcelamelara** [comment]
Totally agreed, and the main purpose of a `Reference` type, as I viewed it, is to provide information to be able to identify and locate an artifact _not_ included in an attestation. I do understand now, though, what you meant by the `value` field potentially being mutually exclusive with the `digest`. I'm not immediately opposed, but I'm also weary of scope creep: in the case of an artifact being included in-line, some of the `Reference` fields like `localName` or `uri` may still be useful, but I think they are mostly not needed because the origins and semantics of the artifact may already be indicated through other information in the `buildDefinition` and `runDetails`. 

@lcarva Are there other fields in the `ArtifactReference`, as currently proposed, that you'd be using in this particular scenario?

**Comment by lcarva** [comment]
> @lcarva Are there other fields in the `ArtifactReference`, as currently proposed, that you'd be using in this particular scenario?


- `uri` makes sense based on @MarkLodato's definition.
- `digest` is a bit redundant if there is a `value`, but I don't have a strong position on whether or not it should be mutually exclusive to `value`. Maybe it shouldn't for consistency sake. This would allow consumers to only care about `value` if they have to. Otherwise, they can just use `digest` as is.
- `localName` would be useful. For Tekton Chains, this could be the name of the result since a Task could emit multiple ones.
- `downloadLocation` would probably just be `uri`.
- `mediaType` is obviously very helpful as it reduces ambiguity when consuming the value.

**Comment by marcelamelara** [comment]
@lcarva Thanks! Would it be possible for you to provide a short JSON example for what a whole ArtifactReference with in-line byproducts might look like for Tekton? I'm in the process of updating the spec for `ResourceDescriptor` (what we were formally calling `ArtifactReference`, and would like to include an example that shows this particular use case.

**Comment by marcelamelara** [comment]
@lcarva PTAL to see if this new field type I'm proposing as part of the in-toto Attestation Framework meets all your requirements: https://github.com/in-toto/attestation/pull/143

**Comment by lcarva** [comment]
@marcelamelara, thanks for putting this together!

Let's say a TaskRun named `check` produces a result named `report` which contains the following value:

```json
{"result":"SUCCESS","timestamp":"1677872236","successes":22,"failures":0,"warnings":0} 
```

The `ResourceDescriptor` could be something like this (note: I'm taking a guess here, this is by no means anything final):

```json
{
  "uri": "https://tekton.dev/TaskRun/check/result/report",
  "digest": {
    "sha256": "ec87961dbfe8e7d8a73890c602ac7bd407b80b9e31c326beb9110bdd255f12e6"
  },
  "representation": {
    "result": "SUCCESS",
    "timestamp": "1677872236",
    "successes": 22,
    "failures": 0, "warnings":0
  },
  "mediaType": "application/json"
}
```

I see that in your proposal `representation` is a JSON object, thus my example above reflects that. This may make it difficult to compute the digest because a JSON object can be formatted in many different ways. It will semantically have the same content, but the digest will be different.

Also, what if I want to include a plain piece of text, or a yaml list? I think the `mediaType` serves this purpose.

Maybe, `representation` should be a string. It is up to consumers to interpret that based on the `mediaType`. WDYT?

**Comment by marcelamelara** [comment]
@lcarva Thank you for the example! Yes, I've gotten more comments about the format of the `representation` field. One suggestion was to have it be a base-64 encoded string, which I think makes perfect sense. This would also cover the plain text and yaml list cases you mention. As you say, the `mediaType` can then be used to indicate to consumers how to parse the `representation` (but it's not a strict requirement as the `uri` might also provide this hint instead).

**Comment by lcarva** [comment]
Nice! base64 makes perfect sense to me.

**Comment by marcelamelara** [comment]
Addressed by #684
